### PR TITLE
WIP: Add more testing + memory stats from cgroups

### LINKED ISSUE #2: Please read memory usage data from cgroups
Similar to https://github.com/povilasv/systemd_exporter/issues/1 but memory related stats:

```
# cd /sys/fs/cgroup/memory/system.slice/system-postgresql.slice
/sys/fs/cgroup/memory/system.slice/system-postgresql.slice# grep . *
cgroup.clone_children:0
grep: cgroup.event_control: Invalid argument
memory.failcnt:0
grep: memory.force_empty: Invalid argument
memory.kmem.failcnt:0
memory.kmem.limit_in_bytes:9223372036854771712
memory.kmem.max_usage_in_bytes:52396032
memory.kmem.tcp.failcnt:0
memory.kmem.tcp.limit_in_bytes:9223372036854771712
memory.kmem.tcp.max_usage_in_bytes:0
memory.kmem.tcp.usage_in_bytes:0
memory.kmem.usage_in_bytes:17018880
memory.limit_in_bytes:9223372036854771712
memory.max_usage_in_bytes:17903652864
memory.move_charge_at_immigrate:0
memory.numa_stat:total=0 N0=0
memory.numa_stat:file=0 N0=0
memory.numa_stat:anon=0 N0=0
memory.numa_stat:unevictable=0 N0=0
memory.numa_stat:hierarchical_total=226104 N0=226104
memory.numa_stat:hierarchical_file=18420 N0=18420
memory.numa_stat:hierarchical_anon=207684 N0=207684
memory.numa_stat:hierarchical_unevictable=0 N0=0
memory.oom_control:oom_kill_disable 0
memory.oom_control:under_oom 0
memory.oom_control:oom_kill 0
grep: memory.pressure_level: Invalid argument
memory.soft_limit_in_bytes:9223372036854771712
memory.stat:cache 0
memory.stat:rss 0
memory.stat:rss_huge 0
memory.stat:shmem 0
memory.stat:mapped_file 0
memory.stat:dirty 0
memory.stat:writeback 0
memory.stat:pgpgin 0
memory.stat:pgpgout 0
memory.stat:pgfault 0
memory.stat:pgmajfault 0
memory.stat:inactive_anon 0
memory.stat:active_anon 0
memory.stat:inactive_file 0
memory.stat:active_file 0
memory.stat:unevictable 0
memory.stat:hierarchical_memory_limit 9223372036854771712
memory.stat:total_cache 921415680
memory.stat:total_rss 7901184
memory.stat:total_rss_huge 0
memory.stat:total_shmem 846168064
memory.stat:total_mapped_file 135168
memory.stat:total_dirty 811008
memory.stat:total_writeback 0
memory.stat:total_pgpgin 18059580
memory.stat:total_pgpgout 17833660
memory.stat:total_pgfault 17232468
memory.stat:total_pgmajfault 0
memory.stat:total_inactive_anon 27570176
memory.stat:total_active_anon 823103488
memory.stat:total_inactive_file 67342336
memory.stat:total_active_file 8105984
memory.stat:total_unevictable 0
memory.swappiness:60
memory.usage_in_bytes:943161344
memory.use_hierarchy:1
notify_on_release:0
grep: postgresql@11-main.service: Is a directory
# 
```

memory.total_rss and memory.usage_in_bytes are probably of biggest interest, but would be great to have almost all of that (especially total_* ones) as metrics.

### PR DESCRIPTION
**Pull request by hamiltont**
PR adds: 

- Reports unit memory metrics by reading `memory.stat` from cgroupfs
- Pulls cgroup code into distinct package to prep for upstreaming
- Improves TravisCI setup - [example](https://travis-ci.org/hamiltont/systemd_exporter)
- - Adds end-to-end integration test for two live systemd versions
- - Adds unit tests
- Increases scrape speed by about 15% by reducing dbus calls 
- Uploads code coverage - [example](https://codecov.io/gh/hamiltont/systemd_exporter/branch/feat%2Fadd-circleci)

Remaining TODOs
 - Decide what memory stats we should export. There are a lot. @povilasv any thoughts you have would be appreciated here. It seems silly to start exporting 20 new metrics per unit so a decision should be made w.r.t. what memory metrics are most impactful 
 - Cleanup docs
 - Cleanup cgroup/cgroup.go file

fixes #2

### CODE REVIEW

**Comment by povilasv** [comment]
Hey :wave:  let me know when it's ready for review :)

**Comment by hamiltont** [comment]
@povilasv Sure, anytime now should be OK. This PR is big already (sorry, seems to be my bad habit), so let's aim to clean it up as needed and get it merged before the next push ;-) I could use any insight you have on "what memory values matter enough to export as metrics"

**Comment by hamiltont** [comment]
Also, a second concern - what should the exported metrics be named? node_exporter moved to using the pattery `node_memory_Capname_metric` (see screenshot). I could not find a good discussion on why they changed their naming pattern, or why they export so many metric names instead of just having something like `node_memory{stat="anon"}`. Regardless, this is a good time to consider how systemd_exporter wants timeseries for memory metrics to be organized

**Comment by povilasv** [comment]
Will review this week thanks for this :)

Re naming I am not sure why node exporter does this but I think we should follow https://prometheus.io/docs/practices/naming/ and do whatever makes sense for our use case and query patterns.

I feel like doing `stat=anon` will make one metric be super big which will suffer from huge cardinality explosion so maybe a metric per `stat`

@SuperQ Would be great to get some insights from you regarding naming :)

**Comment by SuperQ** [comment]
Labels only make sense when there is an aggregation that can be applied. Using a label for `node_memory_...` doesn't make sense because the various values don't add up or average together. For example: `sum(node_memory_bytes)` doesn't make any sense since `MemTotal` + `MemAvailable` bytes is nonsensical.

Having separate metrics also helps for more typical use cases like `node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes`. If it were a label, you would have to do something crazy like `sum without (stat) (node_memory_bytes{stat="MemAvailable"}) / sum without (stat) (node_memory_bytes{stat="MemTotal"})`.

**Comment by hamiltont** [comment]
That's very helpful, thanks

**Comment by baryluk** [comment]
My opinion on metrics selection. I do not see harm on essentially exporting all of them. If they are constant, they will be well compressed by Prometheus and consume almost no resources.

The more interesting question would be which metrics to keep as a separate metric/gauge, and which ones to put together with separate labels. I think it is useful to have some metrics grouped into single gauge with separate labels, if it does make sense to 1) visualise all of them at the same time, 2) make sum of them to see for a total of something.

From a quick look it looks like this only applies to few metrics.

1) rss and rss_huge could be put together. Unless rss from cgroups already accounts for huge allocations.

2) faults. it might make sense to use single counter with minor and major label values. But I do expect major to be what most people look at, with minor providing almost no information, and being significantly bigger often, leading to hiding of actual information. So that might still be beneficial to keep them separate.

As of the others metrics, I think they should be all separate (because doing aggregation over them doesn't make much sense), but I didn't look too deeply into this, and there might be some exceptions.

**Comment by SuperQ** [comment]
Thanks @baryluk

One thing to point out, is if we have metrics that have parts where they can be summed, and we also have a "total" exposed as another part, we try and avoid exposing those in Prometheus best practices. This allows cleaner use of `sum(metric)`.

**Comment by baryluk** [comment]
> Thanks @baryluk
> 
> One thing to point out, is if we have metrics that have parts where they can be summed, and we also have a "total" exposed as another part, we try and avoid exposing those in Prometheus best practices. This allows cleaner use of `sum(metric)`.

Sounds good!

Following Prometheus best practices is a very good idea, to reduce amount of user confusion, but in general it is actually very logical, short and well written to avoid common mistakes with naming metrics and labels.

However regarding your comment, I can't find a general guideline in Prometheues best practices about not exporting `total` metric if `sum(other_metric)` provides same information (modulo minor atomicity differences and moving the aggregation cost from one process to another, and adding a bit extra of memory / storage).

I checked https://prometheus.io/docs/practices/naming/

File: main.go [hunk 1]
```diff
@@ -85,8 +122,5 @@ func main() {
 		}
 	})
 
-	log.Infoln("Listening on", *listenAddress)
-	if err := http.ListenAndServe(*listenAddress, nil); err != nil {
-		log.Fatal(err)
-	}
+	return *listenAddress
```
Comment (povilasv) [inline main.go]: Let's make this return custom https://golang.org/pkg/net/http/#NewServeMux instead of listen address

File: main.go [hunk 2]
```diff
@@ -14,6 +16,40 @@ import (
...
+
+		// Reset http package
+		http.DefaultServeMux = http.NewServeMux()
+		log.Infoln("Test server shutdown")
+	}()
+
+	return srv
+}
+
+func mainCore() string {
```
Comment (povilasv) [inline main.go]: I feel like this needs a better name :) But not sure what to suggest, maybe run?

File: main.go [hunk 3]
```diff
@@ -37,6 +73,7 @@ func main() {
 	kingpin.Version(prommod.Print(version.Print("systemd_exporter")))
 	kingpin.HelpFlag.Short('h')
 	kingpin.Parse()
+	log.Debugf("Parsed '%s'", os.Args)
```
Comment (povilasv) [inline main.go]:
```suggestion
	log.Debugf("Parsed '%s\n'", os.Args)
```

File: main.go [hunk 4]
```diff
@@ -14,6 +16,40 @@ import (
...
+	listenAddress := mainCore()
+
+	log.Infoln("Listening on", listenAddress)
+	if err := http.ListenAndServe(listenAddress, nil); err != nil {
+		log.Fatal(err)
+	}
+
+}
+
+func testMain(wg *sync.WaitGroup) *http.Server {
```
Comment (povilasv) [inline main.go]: Consider moving this into testing package.

File: main_test.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	// b, err := ioutil.ReadAll(resp.Body)
+	// if err != nil {
+	// 	return nil, err
+	// }
+	// if err := resp.Body.Close(); err != nil {
+	// 	return nil, err
+	// }
+	// if want, have := http.StatusOK, resp.StatusCode; want != have {
+	// 	return nil, fmt.Errorf("want /metrics status code %d, have %d. Body:\n%s", want, have, b)
+	// }
```
Comment (povilasv) [inline main_test.go]: Let's remove commented out code

File: systemd/systemd_test.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+		Path:        "",
+		JobId:       0,
+		JobType:     "",
+		JobPath:     "",
+	}
+	found := parseUnitType(x)
+	if found != "service" {
+		t.Errorf("Bad unit name parsing. Wanted %s got %s", "service", found)
+	}
+
```
Comment (povilasv) [inline systemd/systemd_test.go]: Let's get rid of empty space

File: systemd/systemd.go [hunk 7]
```diff
@@ -260,80 +301,78 @@ func (c *Collector) collectUnit(conn *dbus.Conn, ch chan<- prometheus.Metric, un
...
+		cgroupPath, err := c.getControlGroup(conn, unit)
+		if err != nil {
+			logger.Warnf(errUnitMetricsMsg, err)
+		}
+		// Everything below requires a cgroup
+		if cgroupPath == nil {
+			break
+		}
+		err = c.collectUnitCPUMetrics(*cgroupPath, conn, ch, unit)
+		if err != nil {
```
Comment (povilasv) [inline systemd/systemd.go]:
```suggestion
		if err != nil &&  parseUnitType(unit) != "socket"  {
```

File: systemd/systemd.go [hunk 8]
```diff
@@ -260,80 +301,78 @@ func (c *Collector) collectUnit(conn *dbus.Conn, ch chan<- prometheus.Metric, un
...
+			// Most sockets do not have a cpu cgroupfs entry, but a few big ones do (notably docker.socket). Quiet down
+			// error reporting if error came from a socket
+			if parseUnitType(unit) != "socket" {
```
Comment (povilasv) [inline systemd/systemd.go]:
```suggestion

```

**Comment by SuperQ** [comment]
@baryluk It's under the labels section of "writing exporters".

https://prometheus.io/docs/instrumenting/writing_exporters/#labels

**Comment by hamiltont** [comment]
Retitled as WIP while I work through the feedback

**Comment by hectorhuertas** [comment]
@hamiltont thanks for all the amazing work you've done here. do you have plans to resume work on this PR? those cgroup rss metrics are very appealing :)

**Comment by hamiltont** [comment]
Thanks! Yes, I actually have some nice commits to push, been busy with my
day job. I'll try to get out an update this week

**Comment by talyz** [comment]
What's the status here? I've been trying this out for a while and it really makes the metrics much more useful. I had to add a small patch to ignore services with the `RemainAfterExit` property (https://github.com/talyz/systemd_exporter/commit/284e372b8dbe12b29031ac0d19c551167cc75fff) in order for it to not spam my logs, but that's the only issue I've had with it so far.

**Comment by hamiltont** [comment]
Well unfortunately real life jumped up and took hold, but I am thrilled to
hear that some folks found it useful

At this point.... If anyone has time to wrap this work up (as I recall
that's mainly responding to the feedback that was given) it would be super
appreciated to have the help

I'll see if I can push all the other improvements that I had to a distinct
branch in case anyone is interested. As I recall they were definitely not
ready (I was/am learning golang in the evenings, so while I can see the
concepts needed to make a system better it takes me quite a while to
translate those into useful go code)

**Comment by talyz** [comment]
I see :) Yes, it's really useful - great work!

Unfortunately, I don't think I can be of much help - I only know golang at a very basic level. I hope someone can, though.

**Comment by SuperQ** [comment]
After a very long time, we've moved this to prometheus-community. If you're still interested in working on this, please rebase this change against the latest commits.

**Comment by oseiberts11** [comment]
I tried to "rebase this change against the latest commits." but there were a lot of merging problems, but the result as it is (squashed, because there were simply way too many conflicts in the commit-by-commit rebase) is in #66 .

**Pull request closed without merging**